docs/docs/guides/03-typescript.md

```diff
@@ -2,4 +2,226 @@
+// service.d.ts
+export type Admin = Record<string, any>;
+
+// uid.d.ts
+export type Admin = `admin::${string}`;
+
+// index.d.ts
+export * as Service from './service';
+export * as UID from './uid';
+
+// src/index.ts
+import type { Service, UID } from './index';
+
+type AdminService = Record<UID.Admin, Service.Admin>;
+```
+
+### b. Imports
+
+When importing **only types** from a module, `import type` must be used.
+
+The `import type` syntax imports only the types and eliminates the runtime impact of importing the actual values. This helps reduce the bundle size and improves performance by avoiding unnecessary code execution.
+
+Example:
+
+```typescript
+// ❌ don't
+
+import { Attribute, Common } from '@strapi/strapi';
+
+// ✅ do
+
+import type { Attribute, Common } from '@strapi/strapi';
+```
+
+When importing **both types and implementations** from a module:
+
+- `import type` must be used for types
+- `import` must be used for the rest
+
+Example:
+
+```typescript
+// ❌ don't
+
+import { User, findUser } from './user';
+
+const user: User = findUser();
+
+// ✅ do
+
+import type { User } from './user';
+import { findUser } from './user';
+
+const user: User = findUser();
+```
+
+### c. Exports
+
+In most cases, it is recommended to use regular exports (export) instead of the default export (export default). Regular exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.
+
+However, there are a few exceptions where using export default might be appropriate. For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.
+
+Using regular exports provides a clearer and more explicit interface for module consumers, promoting better code organization and maintainability. However, it is important to consider the context before deciding whether to use default exports.
+
+## 3. Generic Types' Parameters
+
+When working with generic types, it is required to follow naming conventions and use descriptive names for type parameters. This helps improve code readability and maintainability over the whole codebase.
+
+- For simple types with a single self-explanatory parameter, it is allowed to **use `T`**
+- For any other generic types' parameters, it is required to **use a meaningful PascalCased variable name prefixed by `T`**
+
+Example:
+
+```typescript
+// ❌ don't
+function query<T extends object, U extends keyof T>(data: T, properties: U[]): { [K in U]: T[K] } {
+  // ...
+}
+
+// ✅ do
+function query<TValue extends object, TKey extends keyof TValue>(
+  data: TValue,
+  properties: TKey[]
+): { [TIter in TKey]: TValue[TIter] } {
+  // ...
+}
+
+// ❌ avoid
+type NonEmpty<TValue extends string> = TValue extends '' ? never : TValue;
+
+// ✅ prefer
+type NonEmpty<T extends string> = T extends '' ? never : T;
```

Comment:
I thought we always prefered verbose, we just had an exception to allow single letters if it was self-explanatory.

Reply:
Don't you consider this to be self-explanatory? (between the name of the type, the extends string & the fact that the type definition is quite straight forward)

Reply:
(I swear I replied to this but now I don't see it)

I mean that I thought we had discussed both options would be ✅ rather than avoiding the first one. If someone wants to write it out, that's just as acceptable, but they don't need to if it's that simple.